Handle Dates-array arithmetic with promote_op

timholy opened the pull request:
This fixes the ambiguity warnings triggered by #12115. It also introduces a new abstract type, `AbstractScalar` and makes use of it in defining operations.

CC @GordStephen, @quinnj.

tkelman commented:
Can you describe briefly what the difference is between `Number` and `AbstractScalar` ? Which of `Number`'s behaviors are not going to be defined by default for `AbstractScalar` ? Would any promotion (or other) documentation need to be adjusted for this?

quinnj commented on base/dates/types.jl:
```diff
@@ -1,6 +1,6 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-abstract AbstractTime
+abstract AbstractTime <: AbstractScalar
```
Hmmm... I guess this is ok. AbstractTime was kind of a catch all for all the Dates types, but this should be ok. The piece that may not make as much is the parsing Slot types, but those probably need to be reconfigured anyway.

timholy commented:
Basically, the idea is that if someone was reluctant to call their type a `Number`, we should have some notion of `AbstractScalar`. So this is perhaps more of a question for @quinnj about why it isn't a `Number`.

andreasnoack commented:
I've actually wanted something like an `AbstractScalar` or `NonArray` for a while. Right now we have [this definition](https://github.com/JuliaLang/julia/blob/d7351cf16fbcd89ae0cf3bac58e045a87ea00c11/base/operators.jl#L152) for `transpose` which I'd really like to have narrowed down to something like `NonArray`/`AbstractScalar`

timholy commented:
I should add that another approach would be to use traits and introduce `isscalar`.

tkelman commented:
This question comes up a lot, a few times at JuliaCon for example, people often want to have abstract container types that are number-like but aren't necessarily compatible with every single thing that we define in base for `Number`. Maybe one more level of granularity in the abstract type hierarchy wouldn't hurt, but the real solution to the "which behaviors do you want" problem is out of scope here and will probably be something traits-like.

timholy commented:
If we want to define `AbstractTime <: Number` then I'd be happy indeed to drop the first commit.

lindahua commented:
*scalar* (in math context) is mainly used to refer to the coefficients in linear algebra (see https://en.wikipedia.org/wiki/Scalar_(mathematics)). In this sense, it might be more specialized (instead of more general) than *number*.

What about call it `Atom`, which indicates that the instance should be considered as a whole, instead of being a collection of ingredients.

lindahua commented:
In the mean time, for symmetry, we may also consider introducing something like `AbstractContainer`.

mbauman commented:
Some (brief) prior discussion here: https://github.com/JuliaLang/julia/pull/9270#issuecomment-66224853, which also mentions `Atom`.  I feel like there's been other issues discussing this idea, but I can't find them now.  If we do this, it should probably be exported like the other abstract types.

I don't like how this moves arithmetic from `Number` to `AbstractScalar`.  I think arithmetic is inherently a number-y thing to do.  While time `Period`s could be `Number`s, I don't think `Date` should be.

timholy commented:
Thanks for the excellent discussion.

I'm tempted to rework this using traits. Methods like `*(A::AbstractArray, B::Number) = A .* B` would be redefined as

```jl
abstract IsContainerTrait
immutable AbstractContainer <: IsContainerTrait end
immutable AbstractAtom <: IsContainerTrait end

iscontainertrait(::AbstractArray) = AbstractContainer()
iscontainertrait(::Number) = AbstractAtom()
iscontainertrait(::AbstractTime) = AbstractAtom()

@inline *(A, B) = *((A, iscontainertrait(A)), (B, iscontainertrait(B)))
function *{AT,BT}(Atrait::Tuple{AT,AbstractContainer}, Btrait::Tuple{BT,AbstractAtom})
    ....
end
```

**However**, I worry about adding yet another system (even for internal use) this late in 0.4. At this point, I wonder if the best course would be to revert #12115 and have people who need this functionality get it via a package. See https://github.com/JuliaLang/julia/pull/12115#issuecomment-124548123. Thoughts? Especially @quinnj, who hasn't chimed in about that yet.

mbauman commented:
It seems like we really want to define `*(::AbstractArray, ::Not{AbstractArray})`, with some magical `Not{}` that's the opposite of `Union{}`.

StefanKarpinski commented:
> It seems like we really want to define `*(::AbstractArray, ::Not{AbstractArray})`, with some magical `Not{}` that's the opposite of `Union{}`.

There goes decidability of subtyping.

jakebolewski commented:
You can do subtyping with negation types and have it be decidable.

StefanKarpinski commented:
Sure, but is Julia's subtyping decidable if you add negation types? I'm not sure, but I doubt it.

timholy commented:
OK, I think I found a minimalistic way to do this. When I looked more carefully, the Array/Scalar operations were not really the core source of the problem.

mbauman commented on base/functors.jl:
```diff
@@ -58,6 +58,9 @@ call(::DotMulFun, x, y) = x .* y
 call(::RDivFun, x, y) = x / y
 
+immutable DotRDivFun <: Func{2} end
+call(::RDivFun, x, y) = x ./ y
```
`call(::DotRDivFun`?

timholy replied:
'Doh! Thanks.

timholy committed:
Base AbstractArray{T<:Date} operations on promote_op

This avoids deprecation warnings with packages like Images and DataArrays
(hopefully others too). The key point seems to be to avoid specializing
on the element type while being generic about the container.

tkelman commented:
Yeah, this looks much simpler and uncontroversial now.
